lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -983,41 +1055,303 @@ void crypto_cipher_final(void *ctx __unused, uint32_t algo __unused)
  * Message Authentication Code functions
  *****************************************************************************/
 #if defined(_CFG_CRYPTO_WITH_MAC)
-TEE_Result crypto_mac_alloc_ctx(void **ctx __unused, uint32_t algo __unused)
+static TEE_Result mac_get_ctx_size(uint32_t algo, size_t *size)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		*size = sizeof(mbedtls_md_context_t);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC_MAC)
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		return TEE_ERROR_NOT_SUPPORTED;
+#endif
+#if defined(CFG_CRYPTO_CMAC)
+	case TEE_ALG_AES_CMAC:
+		return TEE_ERROR_NOT_SUPPORTED;
+#endif
+	default:
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return TEE_SUCCESS;
 }
 
-void crypto_mac_free_ctx(void *ctx, uint32_t algo __unused)
+TEE_Result crypto_mac_alloc_ctx(void **ctx_ret, uint32_t algo)
 {
-	if (ctx)
-		assert(0);
+	const mbedtls_md_info_t *md_info __maybe_unused;
+	int lmd_res __maybe_unused;
+	TEE_Result res = TEE_SUCCESS;
+	size_t ctx_size;
+	void *ctx;
+
+	res = mac_get_ctx_size(algo, &ctx_size);
+	if (res)
+		return res;
+
+	ctx = calloc(1, ctx_size);
+	if (!ctx)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		md_info = get_hash_info(algo);
+		if (!md_info) {
+			res =  TEE_ERROR_NOT_SUPPORTED;
+			break;
+		}
+
+		mbedtls_md_init(ctx);
+		lmd_res = mbedtls_md_setup(ctx, md_info, 1);
+		if (lmd_res != 0) {
+			EMSG("md setup failed, res is 0x%x", -lmd_res);
+			res = TEE_ERROR_GENERIC;
+		}
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC_MAC)
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		break;
+#endif
+#if defined(CFG_CRYPTO_CMAC)
+	case TEE_ALG_AES_CMAC:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		break;
+		break;
+#endif
+	default:
+		res = TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	if (res == TEE_SUCCESS)
```

Comment:
I'd prefer doing the epilogue of this function as:
```
    *ctx_ret = ctx;
    return TEE_SUCCESS;
err:
    free(ctx);
    return res;
```
It's more like it's done in the Linux kernel and also at each point where we break out we know if it's because of an error or not so doing "goto err" wouldn't be harder.

Reply:
Ok, we can do like that. But I still want to use `crypto_mac_free_ctx()` for `err`. Because there are may allocate operation in `mbedtls_md_setup()`, it maybe failed in middle process. So we need to call `crypto_mac_free_ctx()` to free allocated memory.

Reply:
So, as discuss above, we should use `free(ctx)` here. 
But I find another issue, caller will call `crypto_mac_free_ctx()` if `crypto_mac_alloc_ctx()` failed. it will cause double free for `ctx`, look this example: 
https://github.com/OP-TEE/optee_os/blob/af8149de7c373fd859175f9989a64915832ee8c0/core/tee/tee_cryp_hkdf.c#L37-L62

Reply:
Calling `crypto_mac_free_ctx()` on a `ctx` being `NULL` is harmless, it's even preferred over testing if `ctx` is `NULL` or not before calling the function.

If I would have written that function, I would have returned on the `crypto_mac_alloc_ctx()` error instead of doing a goto. However, this is a matter of taste so we leave that to the author.

You're free to do the function epilogue as you like, I just think what I proposed instead is less complicated.

Reply:
Ok, got, thanks for you explaination.